src/main/scala/scorex/core/ModifiersCache.scala

```diff
@@ -0,0 +1,143 @@
+
+import scala.collection.mutable
+import scala.util.{Failure, Success}
+
+/**
+  * A cache which is storing persistent modifiers not applied to history yet.
+  * @tparam PMOD - type of a persistent node view modifier (or a family of modifiers).
+  */
+trait ModifiersCache[PMOD <: PersistentNodeViewModifier, H <: HistoryReader[PMOD, _]] {
+  require(maxSize >= 1)
+
+  def size: Int = currentCacheSize
+
+  /**
+    * How many elements are to be stored in the cache
+    */
+  def maxSize: Int
+
+  type K = mutable.WrappedArray[Byte]
+  type V = PMOD
+
+  protected var currentCacheSize: Int = 0
+
+  protected val cache = mutable.Map[K, V]()
+
+  /**
+    * Keys to simulate objects residing a cache. So if key is stored here,
+    * the membership check (contains()) shows that the key is in the cache,
+    * but the value corresponding to the key is not stored. The motivation
+    * to have this structure is to avoid repeatedly downloading modifiers
+    * which are unquestionably invalid.
+    */
+  protected val rememberedKeys = mutable.HashSet[K]()
+
+  /**
+    * Defines a best (and application-specific) candidate to be applied.
+    * @param history - an interface to history which could be needed to define a candiate
+    * @return - candidate if it is found
+    */
+  def findCandidateKey(history: H): Option[K]
+
+  protected def onPut(key: K): Unit = {}
+  protected def onRemove(key: K, rememberKey: Boolean): Unit = {}
+
+  /**
+    * A cache element replacement strategy method, which defines a key to remove from cache when it is overfull
+    */
+  protected def keyToRemove(): K
+
+
+  def contains(key: K): Boolean = cache.contains(key) || rememberedKeys.contains(key)
+
+  def put(key: K, value: V): Unit = synchronized {
+    if(!contains(key)) {
+      onPut(key)
+      cache.put(key, value)
+      currentCacheSize = currentCacheSize + 1
+
+      if (currentCacheSize > maxSize) remove(keyToRemove())
+    }
+  }
+
+  /**
+    * Remove an element from the cache.
+    * @param key - modifier's key
+    * @param rememberKey - whether to remember the key as belonging to cache. E.g. invalid modifiers are
+    *                    to be remembered (for not to be requested from the network again).
+    * @return
+    */
+  def remove(key: K, rememberKey: Boolean = false): Option[V] = synchronized {
+    cache.remove(key).map {removed =>
+      onRemove(key, rememberKey)
+      currentCacheSize = currentCacheSize - 1
+      if (rememberKey) rememberedKeys += key
+      removed
+    }
+  }
+
+  def popCandidate(history: H): Option[V] = synchronized {
+    findCandidateKey(history).flatMap(k => remove(k))
+  }
+}
+
+trait LRUCache[PMOD <: PersistentNodeViewModifier, HR <: HistoryReader[PMOD, _]] extends ModifiersCache[PMOD, HR] {
+
+  type Counter = Long
+
+  private var cnt = 0: Counter
+
+  private val keyIndex = mutable.Map[K, Counter]()
+
+  override protected def onPut(key: K): Unit = {
+    keyIndex.put(key, cnt)
+    cnt = cnt + 1
```

Comment:
This will eventually overflow...

Reply:
And the overflow would suddenly cause an unexpected behaviour.

Reply:
You could solve this problem (and also the "faster search" problem) by putting the keys in a `Queue` with a finite maximum size and popping the head of the queue, instead of putting them in a map with counters and popping the key with the minimum counter.

Reply:
@ceilican the problem here is that we need to remove an arbitrary element from the queue (and this operation is to be used more often when getting the first element), and this operation is about linear time for Queue I guess. So we need for Queue alternative with sub-linear by-key removal and sublinear min-value search. Doable if we have two indexes (Map[K, Count] and Map[Count,K]), but then we are doubling memory consumption. Any smarter options?

Reply:
> however, you must also ensure that queue is not becoming too big if caches overflows do happen very rare

Yes, that's a good point. Not sure how I would solve that.